### PR TITLE
[IGA] Adding topology items and python scripts

### PR DESCRIPTION
**Pull request by tteschemacher**
I am adding here the data structure for the brep geometry topology object for Nurbs-based CAD simulations. 

Additionally, the necessary Python scripts are added to allow the standard way to simulate with IGA similiar to the Structural Mechanics Application

### CODE REVIEW

File: applications/IgaApplication/python_scripts/iga_analysis.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# Importing the base class
+from analysis_stage import AnalysisStage
+
+class IgaAnalysis(AnalysisStage):
+    """
+    This class is the main-script of the IgaApplication put in a class
+
+    It can be imported and used as "black-box"
+    """
+    def __init__(self, model, project_parameters):
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_analysis.py]: you can remove this constructor. I don't think you will use MPI anytime soon ...

File: applications/IgaApplication/python_scripts/iga_solver.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+                reaction_variable_z = KratosMultiphysics.KratosGlobals.GetVariable(reaction_variable_name + "_Z")
+                KratosMultiphysics.VariableUtils().AddDof(dof_variable_z, reaction_variable_z, self.main_model_part)
+            else:
+                self.print_warning_on_rank_zero("auxiliary_reaction_list list", "The variable " + dof_variable_name + "is not a compatible type")
+        self.print_on_rank_zero("::[IgaSolver]:: ", "DOF's ADDED")
+
+    def ImportModelPart(self):
+        """This function imports the ModelPart
+        """
+        self._ImportModelPart(self.main_model_part, self.settings["model_import_settings"])
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_solver.py]: I am quite sure this does not work for you

File: applications/IgaApplication/python_scripts/iga_solver.py [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+    def GetComputingModelPart(self):
+        if not self.main_model_part.HasSubModelPart(self.settings["computing_model_part_name"].GetString()):
+            raise Exception("The ComputingModelPart was not created yet!")
+        return self.main_model_part.GetSubModelPart(self.settings["computing_model_part_name"].GetString())
+
+    def ExportModelPart(self):
+        name_out_file = self.settings["model_import_settings"]["input_filename"].GetString()+".out"
+        file = open(name_out_file + ".mdpa","w")
+        file.close()
+        KratosMultiphysics.ModelPartIO(name_out_file, KratosMultiphysics.IO.WRITE).WriteModelPart(self.main_model_part)
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_solver.py]: does this work in your case?

File: applications/IgaApplication/python_scripts/iga_solver.py [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+            self._iga_solution_strategy = self._create_iga_solution_strategy()
+        return self._iga_solution_strategy
+
+    def import_constitutive_laws(self):
+        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
+        if (materials_filename != ""):
+            # Add constitutive laws and material properties from json file to model parts.
+            material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
+            material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
+            KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_solver.py:219]: Are you using this?

File: applications/IgaApplication/python_scripts/iga_static_solver.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        super(StaticIgaSolver, self).__init__(model, custom_settings)
+        self.print_on_rank_zero("::[StaticIgaSolver]:: ", "Construction finished")
+
+    def Initialize(self):
+        self.print_on_rank_zero("::[StaticIgaSolver]:: ", "Initializing ...")
+        super(StaticIgaSolver, self).Initialize() # The mechanical solver is created here.
+        self.print_on_rank_zero("::[StaticIgaSolver]:: ", "Finished initialization.")
+
+    def Solve(self):
+        super(StaticIgaSolver, self).Solve()
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_static_solver.py]: this fct does nothing and can be removed

File: applications/IgaApplication/python_scripts/iga_static_solver.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    def _create_solution_scheme(self):
+        return KratosMultiphysics.ResidualBasedIncrementalUpdateStaticScheme()
+
+    def _create_iga_solution_strategy(self):
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/iga_static_solver.py]: why is this needed here => already exists in baseclass

File: applications/IgaApplication/python_scripts/python_solvers_wrapper_iga.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+from __future__ import print_function, absolute_import, division #makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+
+import KratosMultiphysics
+
+
+def CreateSolver(model, custom_settings):
```
Comment (philbucher) [inline applications/IgaApplication/python_scripts/python_solvers_wrapper_iga.py]: this file is a copy from StructuralMechanics, you don't have most of the things ...

Also it is not up-to-date with the latest developments

**Review by philbucher — CHANGES_REQUESTED** [review]
I suggest to split this PR into two parts, one for the python-scripts and one for the C-stuff

there is a lot of stuff just copied from StructuralMechanics, we should discuss it in person

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:     BSD License
+//           Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Tobias Teschemacher
+//                   Michael Breitenberger
+//                   Thomas Oberbichler
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp:12]: You can drop my name since I haven't done nothing inside this file and I already have a lot of other stuff to maintain :P

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        else
+            return false;
+    }
+
+    ///Constructor
+    BrepEdge::BrepEdge(
+        int rBrepId,
+        std::vector<EdgeTopology>& rBrepEdgeTopologyVector,
+        std::vector<TrimmingRange>& rTrimmingRangeVector,
+        int& rDegree,
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp]: Maybe recheck `type&` vs `const type&` vs `const type`. Just for code quality.

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:     BSD License
+//           Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Tobias Teschemacher
+//                   Michael Breitenberger
+//                   Thomas Oberbichler
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h:12]: idem

File: applications/IgaApplication/custom_utilities/brep_topology/brep_model.cpp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:    BSD License
+//              Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Tobias Teschemacher
+//                   Michael Breitenberger
+//                   Thomas Oberbichler
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_model.cpp:12]: idem

File: applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+#include <ANurbs/src/Curve.h>
+
+// Project includes
+#include "iga_application.h"
+#include "iga_application_variables.h"
+
+
+namespace Kratos
+{
+    using namespace ANurbs;
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.h]: Maybe dangerous. Everything is imported into the Kratos namespace. You can put the `using` inside the function bodies.

File: applications/IgaApplication/custom_utilities/nurbs_brep_modeler.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:     BSD License
+//           Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Tobias Tescheamacher
+//                   Thomas Oberbichler
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/nurbs_brep_modeler.cpp:11]: idem

File: applications/IgaApplication/iga_application_variables.cpp [hunk 14]
```diff
@@ -13,9 +13,12 @@
 namespace Kratos
 {
 
+KRATOS_CREATE_VARIABLE(double, FACE_BREP_ID)
```
Comment (oberbichler) [inline applications/IgaApplication/iga_application_variables.cpp]: `double` or `int`?

File: applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+          m_embedded_points(rEmbeddedPoints),
+          m_model_part(rModelPart),
+          m_control_points_ids(rControlPointIds),
+          IndexedObject(rBrepId),
+          Flags()
+    {
+        int number_of_nodes_u = rKnotVectorU.size() + rP - 1;
+        int number_of_nodes_v = rKnotVectorV.size() + rQ - 1;
+
+        m_node_surface_geometry_3d = New<NodeSurfaceGeometry3D>(
```
Comment (oberbichler) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp]: Instead of `ANurbs::New` you can use `Kratos::make_unique`. It does exactly the same. You can check out [shared_pointers.h](https://github.com/KratosMultiphysics/Kratos/blob/48be4aef49bab6dc4288d71f64b2044a1c1f3748/kratos/includes/shared_pointers.h#L41) for some helper functions in regard to smart pointers.

File: applications/IgaApplication/python_scripts/iga_solver.py (hunk 4 above)
Comment (tteschemacher) [reply applications/IgaApplication/python_scripts/iga_solver.py:219]: I do. Materials are the same as in the Structural Mechanics

File: applications/IgaApplication/python_scripts/python_solvers_wrapper_iga.py (hunk 7 above)
Comment (tteschemacher) [reply applications/IgaApplication/python_scripts/python_solvers_wrapper_iga.py]: Yes, you are right. That was the wrong version

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp (hunk 8 above)
Comment (tteschemacher) [reply applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp:12]: Haha you can not escape from this.

File: applications/IgaApplication/iga_application_variables.cpp (hunk 14 above)
Comment (tteschemacher) [reply applications/IgaApplication/iga_application_variables.cpp]: changed it

File: applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.h (hunk 12 above)
Comment (tteschemacher) [reply applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.h]: I see the problem. Is there any other option to include Curve2D and CurveGeometry2D?

Comment (oberbichler) [reply applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.h]: Using `ANurbs::Curve2D` or put the `using` inside your function body. Anyway I have to add an alias for these entities into Kratos. After that we only need to change `ANurbs::Curve...` to `Curve...`

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+                {
+                    int id = 0;
+                    if (rModelPart.GetRootModelPart().Conditions().size()>0)
+                        int id = rModelPart.GetRootModelPart().Conditions().back().Id() + 1;
+                    auto condition = rModelPart.CreateNewCondition(rName, id, non_zero_control_points, this_property);
+
+                    condition->SetValue(SHAPE_FUNCTION_VALUES, N_0);
+                    condition->SetValue(SHAPE_FUNCTION_LOCAL_DERIVATIVES, N_1);
+                    condition->SetValue(INTEGRATION_WEIGHT, integration_points[j].weight);
+                }
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp]: @philbucher, do you see the possibility to somehow template the condition and element?

File: applications/IgaApplication/custom_utilities/brep_json_io.h [hunk 17]
```diff
@@ -0,0 +1,71 @@
+#if !defined(KRATOS_BREP_JSON_IO_H_INCLUDED )
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.h:1]: please use the standard-header

File: applications/IgaApplication/custom_utilities/brep_json_io.h [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+// Project includes
+#include "brep_topology/brep_model.h"
+#include "brep_topology/brep_trimming_curve.h"
+#include "brep_topology/brep_boundary_loop.h"
+#include "brep_topology/brep_face.h"
+#include "brep_topology/brep_edge.h"
+
+#include "includes/kratos_parameters.h"
+#include "includes/model_part.h"
+#include "includes/node.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.h:18]: not needed

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+//                   Riccardo Rossi
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "brep_json_io.h"
+#include "iga_application.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: please remove this

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+
+namespace Kratos
+{
+    void BrepJsonIO::WriteIntegrationDomainJson(ModelPart& rModelPart, const std::string& rOutputFileName)
+    {
+        std::ofstream file;
+        file.open(rOutputFileName);
+        std::string separator = "";
+        file << "{\"geometry_integration_points\" :[";
+        for (auto element = rModelPart.ElementsBegin(); element != rModelPart.ElementsEnd(); ++element)
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp:31]: `for (const auto& r_elem : rModelPart.Elements())`

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+        }
+        file << "]}";
+        file.close();
+    }
+
+    std::vector<BrepModel> BrepJsonIO::ImportNurbsBrepGeometry(
+        ModelPart& model_part,
+        Parameters& rNurbsBrepGeometryJson)
+    {
+        std::cout << "\n> Start reading CAD geometry" << std::endl;
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: Use the Kratos-Logger for this

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+            file << separator << "[" << element->Id() << ", " << element->GetValue(FACE_BREP_ID) << ",[" << local_parameters[0] << ", " << local_parameters[1] << "]]";
+            separator = ",";
+        }
+        file << "]}";
+        file.close();
+    }
+
+    std::vector<BrepModel> BrepJsonIO::ImportNurbsBrepGeometry(
+        ModelPart& model_part,
+        Parameters& rNurbsBrepGeometryJson)
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: Parameters should be passed by-value

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+    {
+        std::cout << "\n> Start reading CAD geometry" << std::endl;
+
+        double model_tolerance = rNurbsBrepGeometryJson["tolerances"]["model_tolerance"].GetDouble();
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: const

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+
+        std::vector<BrepModel> r_brep_model_vector;
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: `brep_model_vector`, this is not a reference

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+
+        for (int brep_i = 0; brep_i < rNurbsBrepGeometryJson["breps"].size(); brep_i++)
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp:54]: `std::size_t`

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 26]
```diff
@@ -0,0 +1,454 @@
...
+        {
+            Parameters brep_json = rNurbsBrepGeometryJson["breps"][brep_i];
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp:56]: const auto

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 27]
```diff
@@ -0,0 +1,454 @@
...
+            int brep_brep_id = brep_json["brep_id"].GetInt();
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp]: cpnst

File: applications/IgaApplication/custom_utilities/brep_json_io.cpp [hunk 28]
```diff
@@ -0,0 +1,454 @@
...
+            const Vector& local_parameters = element->GetValue(LOCAL_COORDINATES);
+            //const Vector& global_coordinates = element->Calculate(COORDINATES);
+            file << separator << "[" << element->Id() << ", " << element->GetValue(FACE_BREP_ID) << ",[" << local_parameters[0] << ", " << local_parameters[1] << "]]";
+            separator = ",";
+        }
+        file << "]}";
+        file.close();
+    }
+
+    std::vector<BrepModel> BrepJsonIO::ImportNurbsBrepGeometry(
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.cpp:42]: 2mmfnh

Please refactor this into smaller fcts, this is ~400 lines long!

E.g. into edges, faces etc
Please take into account the comment I made abt other parts of the code

File: applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.cpp [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+namespace Kratos
+{
+    std::vector<BrepTrimmingCurve> BrepBoundaryLoop::GetTrimmingCurves()
+    {
+        return m_brep_trimming_curves;
+    }
+
+    bool& BrepBoundaryLoop::IsOuterLoop()
+    {
+        return m_is_outer_loop;
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.cpp]: members should be `mIsOuterLoop`

File: applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.cpp [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+
+
+namespace Kratos
+{
+    std::vector<BrepTrimmingCurve> BrepBoundaryLoop::GetTrimmingCurves()
+    {
+        return m_brep_trimming_curves;
+    }
+
+    bool& BrepBoundaryLoop::IsOuterLoop()
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.cpp]: why by-ref?

File: applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h [hunk 31]
```diff
@@ -0,0 +1,61 @@
+#if !defined(KRATOS_BREP_BOUNDARY_LOOP_H_INCLUDED )
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h:15]: header

File: applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+#define  KRATOS_BREP_BOUNDARY_LOOP_H_INCLUDED
+
+
+// System includes
+
+// Project includes
+#include "brep_trimming_curve.h"
+
+#include "iga_application.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h]: pls remove this

File: applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+        /// Pointer definition of KratosNurbsBrepApplication
+        KRATOS_CLASS_POINTER_DEFINITION(BrepBoundaryLoop);
+
+        ///@}
+        ///@name Life Cycle 
+        ///@{ 
+
+        std::vector<BrepTrimmingCurve> GetTrimmingCurves();
+        bool& IsOuterLoop();
+        //important function, to be implemented!!
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_boundary_loop.h:52]: ???

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+#include "brep_edge.h"
+
+namespace Kratos
+{
+    /* If edge has more then one brep topology object, then the edge is a coupling
+    *  edge.
+    */
+    bool BrepEdge::IsCouplingEdge()
+    {
+        if (m_brep_edge_topology_vector.size() > 1)
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp]: this can be simplified to:
`return m_brep_edge_topology_vector.size() > 1`

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+
+                    N_0(m) = shape(0, m);
+
+                    N_1(m, 0) = shape(1, m);
+                }
+                if (rType == "element")
+                {
+                    int id = 0;
+                    if (rModelPart.GetRootModelPart().Elements().size()>0)
+                        id = rModelPart.GetRootModelPart().Elements().back().Id() + 1;
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.cpp]: da rollt es mir ja die Fussnägel auf!
`id = rModelPart.GetRootModelPart().NumberOfElements() + 1`

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h [hunk 36]
```diff
@@ -0,0 +1,160 @@
...
+#if !defined(KRATOS_BREP_EDGE_H_INCLUDED )
+#define  KRATOS_BREP_EDGE_H_INCLUDED
+
+// System includes
+
+// Project includes
+#include "iga_application.h"
+#include "iga_application_variables.h"
+
+#include "../node_curve_geometry_3d.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h]: no relative includes pls

File: applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h [hunk 37]
```diff
@@ -0,0 +1,160 @@
...
+//
+
+
+#if !defined(KRATOS_BREP_EDGE_H_INCLUDED )
+#define  KRATOS_BREP_EDGE_H_INCLUDED
+
+// System includes
+
+// Project includes
+#include "iga_application.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_edge.h]: remove pls

File: applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp [hunk 38]
```diff
@@ -0,0 +1,251 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:     BSD License
+//           Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp]: this is no longer IGAStructuralMechanics => please point to the one in core

File: applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp [hunk 39]
```diff
@@ -0,0 +1,251 @@
...
+
+                        N_0[n] = shape(0, indexU, indexV);
+                        N_1(n, 0) = shape(1, indexU, indexV);
+                        N_1(n, 1) = shape(2, indexU, indexV);
+                        N_2(n, 0) = shape(3, indexU, indexV);
+                        N_2(n, 1) = shape(5, indexU, indexV);
+                        N_2(n, 2) = shape(4, indexU, indexV);
+                    }
+
+                    KRATOS_WATCH(N_0)
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp]: leftover debugging output

File: applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp [hunk 40]
```diff
@@ -0,0 +1,251 @@
...
+    void BrepFace::GetGeometryIntegrationTrimmed(ModelPart& rModelPart,
+        const std::string& rType,
+        const std::string& rName,
+        const int& rPropertiesId,
+        const int& rShapeFunctionDerivativesOrder,
+        std::vector<std::string> rVariables)
+    {
+        //auto clipper = ANurbs::Tri();
+
+        Properties::Pointer this_property = rModelPart.pGetProperties(rPropertiesId);
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_face.cpp]: why do you need a pointer?

File: applications/IgaApplication/custom_utilities/brep_topology/brep_model.cpp [hunk 41]
```diff
@@ -0,0 +1,87 @@
...
+//
+
+// Project includes
+#include "brep_model.h"
+
+#include "iga_application.h"
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_model.cpp]: remove

File: applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.cpp [hunk 42]
```diff
@@ -0,0 +1,66 @@
...
+
+// External includes 
+
+// Project includes
+#include "brep_trimming_curve.h"
+
+
+namespace Kratos
+{
+    int& BrepTrimmingCurve::GetTrimIndex()
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_trimming_curve.cpp]: why reference?

File: applications/IgaApplication/custom_utilities/brep_topology/brep_vertex.cpp [hunk 43]
```diff
@@ -0,0 +1,21 @@
+//    |  /           |
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_topology/brep_vertex.cpp:1]: please remove empty files

unless you will fill it later

File: applications/IgaApplication/custom_utilities/brep_json_io.h [hunk 44]
```diff
@@ -0,0 +1,71 @@
...
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of KratosNurbsTestcaseApplication
+    KRATOS_CLASS_POINTER_DEFINITION(BrepJsonIO);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+    void WriteIntegrationDomainJson(
```
Comment (philbucher) [inline applications/IgaApplication/custom_utilities/brep_json_io.h]: I think this could be free functions, no need to put them in a class

**Review by philbucher — COMMENTED** [review]
I started going through the C-code, but it is too much to look at at once

I gave you a bunch of suggestion for cleaning, please apply them in all files even though I did not point out all locations.

BTW I strongly suggest to create C-tests for the BREP-classes, especially considering the upcoming AMatrix-transition

Ofc it would also be good to have tests for other features like your Brep-json reader
There you should think abt cases in which it can fail => throw a meaningful message in those cases => check in the tests if these messages are thrown
I can show you how o do this

**Comment by philbucher** [comment]
@tteschemacher also take a look at the codacy-suggestions, they are necessary

**Comment by oberbichler** [comment]
@tteschemacher Please use cherrypick and don't copy&paste the code from other commits. Otherwise we will get in trouble when merging this thing.

**Comment by philbucher** [comment]
> @tteschemacher Please use cherrypick and don't copy&paste the code from other commits. Otherwise we will get in trouble when merging this thing.

:D

704e645b-d074-4c66-912f-ba48cff8b769

**Comment by philbucher** [comment]
@tteschemacher I pushed some cleanup
this includes #3414 which should be merged before

**Comment by philbucher** [comment]
I added some more stuff to the python-scripts
@tteschemacher there are some things that should be updated, we should discuss it tmr

**Comment by philbucher** [comment]
this PR is really messy now, it won't be merged, separate PRs will be made

**Comment by philbucher** [comment]
this can be closed now right?

**Comment by loumalouomega** [comment]
@tteschemacher can we delete the branch?

**Comment by tteschemacher** [comment]
Do we need to?

**Comment by loumalouomega** [comment]
> Do we need to?

If it is unused yes, if you are still using it not. I am just asking. But there are many stale branches, and we should clean up a bit or cloning Kratos will become a nightmare.
